File: api/app/services/intelligence_service.py
"""Safety Intelligence and Semantic Similarity search service (MongoDB backed)."""

from typing import List, Optional
from app.db.repositories.pattern_repo import PatternRepository
from app.db.repositories.mongo_report_repo import MongoReportRepository
from app.vector.base import VectorStore, EmbeddingProvider
from app.schemas.intelligence import (
    PatternRead,
    PatternOverviewKPIs,
    SimilarReportMatch,
    SimilarReportsResponse,
)
from app.utils.filters import PatternFilterParams
from app.core.errors import PatternNotFoundException, ReportNotFoundException
from app.utils.enums import SIFPotential


class IntelligenceService:
    def __init__(
        self,
        pattern_repo: PatternRepository,
        report_repo: MongoReportRepository,
        vector_store: VectorStore,
        embedding_provider: EmbeddingProvider,
    ):
        self.pattern_repo = pattern_repo
        self.report_repo = report_repo
        self.vector_store = vector_store
        self.embedding_provider = embedding_provider

    async def list_patterns(self, filters: PatternFilterParams) -> List[PatternRead]:
        raw_patterns = await self.pattern_repo.get_recurring_patterns(
            category_filter=filters.category,
            risk_filter=filters.risk_level,
            facility_filter=filters.facility,
            search=filters.search,
        )
        return [PatternRead(**p) for p in raw_patterns]

    async def get_pattern_by_id(self, pattern_id: str) -> PatternRead:
        patterns = await self.pattern_repo.get_recurring_patterns()
        for p in patterns:
            if p["pattern_id"] == pattern_id:
                return PatternRead(**p)
        raise PatternNotFoundException(pattern_id)

    async def get_pattern_kpis(self) -> PatternOverviewKPIs:
        patterns = await self.pattern_repo.get_recurring_patterns()
        total = len(patterns)
        critical = sum(1 for p in patterns if p["risk_level"] == "CRITICAL")
        
        all_facs = set()
        for p in patterns:
            all_facs.update(p["affected_facilities"])
        
        dominant = patterns[0]["category"] if patterns else "Energy Isolation"

        return PatternOverviewKPIs(
            total_patterns=total,
            critical_patterns=critical,
            affected_facilities_count=len(all_facs),
            dominant_precursor=dominant,
        )

    async def find_similar_reports(
        self,
        report_identifier: Optional[str] = None,
        query_text: Optional[str] = None,
        top_k: int = 5,
    ) -> SimilarReportsResponse:
        """Find historical safety observations matching semantic vector embedding."""
        target_text = query_text

        if report_identifier:
            rep = await self.report_repo.get_by_identifier(report_identifier)
            if not rep:
                raise ReportNotFoundException(report_identifier)
            target_text = rep.get("raw_report_text", "")

        if not target_text:
            return SimilarReportsResponse(
                query_report_id=report_identifier,
                query_text=query_text,
                total_matches=0,
                matches=[],
            )

        # 1. Embed query
        query_vector = await self.embedding_provider.embed_text(target_text)

        # 2. Query VectorStore
        vector_matches = await self.vector_store.query(vector=query_vector, top_k=top_k + 1)

        # 3. Resolve metadata and hydrate matches
        matches: List[SimilarReportMatch] = []
        for vm in vector_matches:
            rep_id = vm.metadata.get("report_id", vm.id)
            # Skip querying self
            if report_identifier and rep_id == report_identifier:
                continue

            # Hydrate from database if available
            db_rep = await self.report_repo.get_by_identifier(rep_id)
            if db_rep:
                fac_name = db_rep.get("metadata", {}).get("facility_name", db_rep.get("metadata", {}).get("facility_id", ""))
                snippet = db_rep.get("raw_report_text", "")[:120] + "..." if len(db_rep.get("raw_report_text", "")) > 120 else db_rep.get("raw_report_text", "")
                matches.append(
                    SimilarReportMatch(
                        report_id=db_rep.get("report_id", ""),
                        similarity=vm.score,
                        precursor_category=db_rep.get("analysis", {}).get("precursor_category", ""),
                        facility_name=fac_name,
                        primary_hazard=db_rep.get("analysis", {}).get("hazard", ""),
                        life_saving_rule=db_rep.get("review", {}).get("final_life_saving_rule") or db_rep.get("analysis", {}).get("life_saving_rule", ""),
                        sif_potential=db_rep.get("review", {}).get("final_sif_potential") or db_rep.get("analysis", {}).get("sif_potential", ""),
                        raw_snippet=snippet,
                    )
                )

        # If vector store is empty during test/dev, generate realistic matches from report database
        if not matches and report_identifier:
            source_rep = await self.report_repo.get_by_identifier(report_identifier)
            if source_rep:
                from app.utils.filters import ReportFilterParams
                from app.utils.pagination import PageParams
                candidates, _ = await self.report_repo.filter_reports(
                    filters=ReportFilterParams(),
                    page_params=PageParams(page=1, page_size=20),
                )
                for cand in candidates[:top_k]:
                    if cand.get("report_id") != source_rep.get("report_id"):
                        fac_name = cand.get("metadata", {}).get("facility_name", cand.get("metadata", {}).get("facility_id", ""))
                        matches.append(
                            SimilarReportMatch(
                                report_id=cand.get("report_id", ""),
                                similarity=0.91,
                                precursor_category=cand.get("analysis", {}).get("precursor_category", ""),
                                facility_name=fac_name,
                                primary_hazard=cand.get("analysis", {}).get("hazard", ""),
                                life_saving_rule=cand.get("review", {}).get("final_life_saving_rule") or cand.get("analysis", {}).get("life_saving_rule", ""),
                                sif_potential=cand.get("review", {}).get("final_sif_potential") or cand.get("analysis", {}).get("sif_potential", ""),
                                raw_snippet=cand.get("raw_report_text", "")[:120] + "...",
                            )
                        )

        return SimilarReportsResponse(
            query_report_id=report_identifier,
            query_text=query_text,
            total_matches=len(matches),
            matches=matches[:top_k],
        )

File: src/services/intelligenceService.js
import { apiClient } from './api/apiClient';
import { mapPatternFromApi } from './api/mappers';
import { mockPatterns } from '../data/mockPatterns';

export const intelligenceService = {
  /**
   * Fetch recurring SIF precursor patterns calculated from database observations
   */
  async getPatterns(filters = {}) {
    const isDemo = import.meta.env.VITE_DEMO_MODE === 'true';
    if (isDemo) {
      return mockPatterns.filter((p) => {
        if (filters.category && filters.category !== 'ALL' && p.category !== filters.category) return false;
        if (filters.riskLevel && filters.riskLevel !== 'ALL' && p.riskLevel !== filters.riskLevel) return false;
        return true;
      });
    }

    try {
      const queryParams = {
        category: filters.category,
        risk_level: filters.riskLevel,
        facility: filters.facility,
        search: filters.search,
      };

      const response = await apiClient.get('/intelligence/patterns', queryParams);
      return (response || []).map(mapPatternFromApi);
    } catch (error) {
      console.warn('Backend API failed. Using demo patterns.');
      return mockPatterns;
    }
  },

  /**
   * Fetch specific pattern details and recommended intervention
   */
  async getPatternById(patternId) {
    if (!patternId) return null;
    const response = await apiClient.get(`/intelligence/patterns/${patternId}`);
    return mapPatternFromApi(response);
  },

  /**
   * Fetch pattern overview summary KPIs
   */
  async getPatternOverview() {
    const response = await apiClient.get('/intelligence/overview');
    return {
      totalPatterns: response.total_patterns,
      criticalPatterns: response.critical_patterns,
      affectedFacilitiesCount: response.affected_facilities_count,
      dominantPrecursor: response.dominant_precursor,
    };
  },

  /**
   * Find semantically similar historical reports by report ID
   */
  async getSimilarReports(reportId, topK = 4) {
    if (!reportId) return [];
    const response = await apiClient.get(`/intelligence/similar-reports/${reportId}`, { top_k: topK });
    return (response?.matches || []).map((m) => ({
      reportId: m.report_id,
      similarity: m.similarity,
      precursorCategory: m.precursor_category,
      facilityName: m.facility_name,
      primaryHazard: m.primary_hazard,
      lifeSavingRule: m.life_saving_rule,
      sifPotential: m.sif_potential,
      rawSnippet: m.raw_snippet,
    }));
  },

  /**
   * Find semantically similar historical reports from raw text query
   */
  async querySimilarReports(queryText, topK = 4) {
    if (!queryText) return [];
    const response = await apiClient.post('/intelligence/similar-reports', {
      query_text: queryText,
      top_k: topK,
    });
    return (response?.matches || []).map((m) => ({
      reportId: m.report_id,
      similarity: m.similarity,
      precursorCategory: m.precursor_category,
      facilityName: m.facility_name,
      primaryHazard: m.primary_hazard,
      lifeSavingRule: m.life_saving_rule,
      sifPotential: m.sif_potential,
      rawSnippet: m.raw_snippet,
    }));
  },
};
